### ISSUE TITLE
Windows Directory containing Chinese will crash

### ISSUE THREAD

**Issue by sunjianan9900** [isar/isar#527]
Windows Directory containing Chinese will crash

**Crashed**
```
 final dir = await getApplicationSupportDirectory();
    debugPrint("isar Star - 2 $dir");
    await Isar.open(
      [ITaskSchema],
      directory: dir.path,
      inspector: true,
    );
```

```
flutter: isar Star - 2 Directory: 'C:\Users\王勇\AppData\Roaming\com.example\demo'
thread 'isarworker' panicked at 'called `Result::unwrap()` on an `Err` value: Utf8Error { valid_up_to: 4, error_len: Some(1) }', D:\a\isar-core\isar-core\src\mdbx\mod.rs:52:39
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
../../third_party/dart/runtime/vm/thread_pool.cc: 315: error: Could not start worker thread: result = 13.
version=2.17.6 (stable) (Tue Jul 12 12:54:37 2022 +0200) on "windows_x64"
Lost connection to device.
pid=11860, thread=2864, isolate_group=vm-service(000001E2377DE7C0), isolate=(nil)(0000000000000
```

**Version**
```
Flutter assets will be downloaded from https://storage.flutter-io.cn. Make sure you trust this source!
Doctor summary (to see all details, run flutter doctor -v):
[√] Flutter (Channel stable, 3.0.5, on Microsoft Windows [版本 10.0.19044.1826], locale zh-CN)
[X] Android toolchain - develop for Android devices
    X Unable to locate Android SDK.
      Install Android Studio from: https://developer.android.com/studio/index.html
      On first launch it will assist you in installing the Android SDK components.
      (or visit https://flutter.dev/docs/get-started/install/windows#android-setup for detailed instructions).
      If the Android SDK has been installed to a custom location, please use
      `flutter config --android-sdk` to update to that location.

[√] Chrome - develop for the web
[√] Visual Studio - develop for Windows (Visual Studio Community 2022 17.2.6)
[!] Android Studio (not installed)
[√] Connected device (3 available)
[√] HTTP Host Availability

! Doctor found issues in 2 categories.
```

isar: ^3.0.0-dev.8
 
isar_flutter_libs: ^3.0.0-dev.8

 isar_generator: ^3.0.0-dev.8

**Comment by leisim** [comment]
I checked the code and your error happened during reading the actual error from the OS. So something else is wrong.

But the crash still needs to be fixed.

**Comment by laohyx** [comment]
Same issue, on windows with Chinese characters.


![image](https://user-images.githubusercontent.com/1221508/181668251-998b50f7-39b8-43ca-876b-1e1af0df0500.png)

**Comment by sunjianan9900** [comment]
> I checked the code and your error happened during reading the actual error from the OS. So something else is wrong.
> 
> But the crash still needs to be fixed.

It looks like isar-core\src\mdbx\mod.rs does not support UTF-8
But we can't ask users to use English names as windows login names because I haven't even found out how to change my login name

**Comment by leisim** [comment]
> It looks like isar-core\src\mdbx\mod.rs does not support UTF-8

Windows does not use UTF8 that is the problem but I need to fix this somehow 🤨

**Comment by laohyx** [comment]
hi @leisim , could you please provide the document about compiling isar-core on Windows?

So that we can help debug on Windows. 

Thanks!

**Comment by leisim** [comment]
1. Install Rust
2. Clone [isar-core](https://github.com/isar/isar-core)
3. Run `cargo build` in the dart-ffi subdirectory.

**Comment by laohyx** [comment]
Hi @leisim , here is the debug anlysis for this issue.

The root cause is `libmdbx` on Windows is using [mbstowcs](https://en.cppreference.com/w/cpp/string/multibyte/mbstowcs) function to convert `char` to `wchar_t` for path string. But `mbstowcs` depends on `locale` to be currectly set. 

When passing the path with non-ASCII chars to mdbx, it converts to the wrong wide char:
![image](https://user-images.githubusercontent.com/1221508/182014354-93b129a6-54cf-4763-9a82-920030b9ea08.png)

Given the bug, when a dirctory name contains non-ASCII, the OS return error code 3, which means the file is not found. 
When calling `ffi::mdbx_strerror`, it returns "系统找不到指定的路径。" in GBK encoding (Chinese system's code page), then the rust raises Utf8Error.
![image](https://user-images.githubusercontent.com/1221508/182014548-7405e0c1-e3b2-47a4-bfc5-4bef49eb509a.png)



Temporay solution: https://github.com/isar/libmdbx/pull/1. Let's set the locale to support UTF-8, and ensure char to wchar conversion. This issue could be solved.

In a pure C/C++ project, I can call `setlocale(LC_ALL, ".utf8");` in the main function and it works. Unfortunately it doesn't work in Flutter. I suspect that the flutter engine overrides the locale somehow. So I hacked `libmbdx` as a temporary solution.


@sunjianan9900 I packaged the fixed version on repo: https://github.com/huaji-app/isar

We are using this pubspec.yaml on production:
```yaml
dependencies:
  # 3.0.0-dev.8 with fixes on https://github.com/isar/isar/issues/527
  isar:
    git:
      url: https://github.com/huaji-app/isar.git
      ref: huaji
      path: packges/isar
  isar_flutter_libs: # contains Isar Core
    git:
      url: https://github.com/huaji-app/isar.git
      ref: huaji
      path: packges/isar_flutter_libs
```

**Comment by MrCsabaToth** [comment]
So is this a bug with libmdbx upstream?

**Comment by leisim** [comment]
@MrCsabaToth yes but I reported it and it has been fixed. waiting for a new stable version.

**Comment by MrCsabaToth** [comment]
@leisim That's very promising!

**Comment by leisim** [comment]
Would you mind testing this again with `3.0.0-dev.9`? Unfortunately I don't have a Chinese Windows device
